src/Command/PurgeCommand.php

```diff
@@ -0,0 +1,249 @@
+    {
+        if (null === $fs) {
+            $fs = new Filesystem();
+        }
+
+        $this->framework = $framework;
+        $this->db = $db;
+        $this->fs = $fs;
+
+        parent::__construct();
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function configure()
+    {
+        $this
+            ->setName('leads:purge')
+            ->setDescription('Purge the leads outsite the configured storage period.');
+    }
+
+    /**
+     * @param InputInterface $input
+     * @param OutputInterface $output
+     * @return int|null|void
+     * @throws \Doctrine\DBAL\DBALException
+     */
+    protected function execute(InputInterface $input, OutputInterface $output)
+    {
+        $this->framework->initialize();
+
+        $purged = $this->executeBatchPurge();
+
+        if ($purged) {
+            $logMessage = 'The leads have been purged successfully.';
+        } else {
+            $logMessage = 'No leads to purge.';
+        }
+
+        $logLevel = LogLevel::INFO;
+        $logger = System::getContainer()->get('monolog.logger.contao');
+        $logger->log($logLevel, $logMessage, array('contao' => new ContaoContext(__METHOD__, $logLevel)));
+        $output->writeln('<info>'.$logMessage.'</info>');
+    }
+
+    /**
+     * @return bool
+     * @throws \Doctrine\DBAL\DBALException
+     */
+    private function executeBatchPurge()
+    {
+        $purged = false;
+        $forms = $this->db->fetchAll("SELECT id, title, leadPeriod FROM tl_form WHERE leadPeriod > 0");
+
+        foreach ($forms as $masterForm) {
+
+            if (!empty($leads = $this->getAllLeads($masterForm))) {
+                $leadsIds = implode(',', array_keys($leads));
+
+                $deletedUploads = null;
+                if (!empty($leadsData = $this->getAllLeadsData($masterForm, $leadsIds))) {
+                    $leadsDataIds = implode(',', array_keys($leadsData));
+                    $deletedData = $this->db->executeUpdate(
+                        "DELETE FROM tl_lead_data WHERE id IN(".$leadsDataIds.")"
+                    );
+
+                    if ($masterForm['leadPurgeUploads']) {
+                        $deletedUploads = $this->purgeUploads($leadsData);
+                    }
+                }
+
+                $deletedLeads = $this->db->executeUpdate(
+                    "DELETE FROM tl_lead WHERE id IN(".$leadsIds.")"
+                );
+
+                $logLevel = LogLevel::INFO;
+                $logMessage = 'Purged leads for master form "'.$masterForm['title'].'": '.(int)$deletedLeads.' leads | '.(int)$deletedData.' data';
+                $logger = System::getContainer()->get('monolog.logger.contao');
+                $logger->log($logLevel, $logMessage, array('contao' => new ContaoContext(__METHOD__, $logLevel)));
+
+                // Add custom logic
+                if (isset($GLOBALS['TL_HOOKS']['postLeadsPurge']) && is_array($GLOBALS['TL_HOOKS']['postLeadsPurge'])) {
```

Comment:
This is new code, it should never contain hooks anymore. Use the event dispatcher for this 😄